File: WebCaster/WebCaster/Services/Storage/PersistenceController.swift
import Foundation

final class PersistenceController {
    static let shared = PersistenceController()

    private let fileManager = FileManager.default
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private var documentsDirectory: URL {
        fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
    }

    private init() {
        encoder.dateEncodingStrategy = .iso8601
        decoder.dateDecodingStrategy = .iso8601
    }

    // MARK: - History

    private var historyFileURL: URL { documentsDirectory.appendingPathComponent("history.json") }

    func loadHistory() -> [HistoryEntry] {
        load(from: historyFileURL) ?? []
    }

    func saveHistory(_ entries: [HistoryEntry]) {
        save(entries, to: historyFileURL)
    }

    func addHistoryEntry(_ entry: HistoryEntry) {
        var history = loadHistory()
        history.insert(entry, at: 0)
        if history.count > 500 { history = Array(history.prefix(500)) }
        saveHistory(history)
    }

    func clearHistory() {
        saveHistory([])
    }

    // MARK: - Bookmarks

    private var bookmarksFileURL: URL { documentsDirectory.appendingPathComponent("bookmarks.json") }

    func loadBookmarks() -> [Bookmark] {
        load(from: bookmarksFileURL) ?? []
    }

    func saveBookmarks(_ bookmarks: [Bookmark]) {
        save(bookmarks, to: bookmarksFileURL)
    }

    func addBookmark(_ bookmark: Bookmark) {
        var bookmarks = loadBookmarks()
        bookmarks.insert(bookmark, at: 0)
        saveBookmarks(bookmarks)
    }

    func removeBookmark(_ bookmark: Bookmark) {
        var bookmarks = loadBookmarks()
        bookmarks.removeAll { $0.id == bookmark.id }
        saveBookmarks(bookmarks)
    }

    // MARK: - Playlists

    private var playlistsFileURL: URL { documentsDirectory.appendingPathComponent("playlists.json") }

    func loadPlaylists() -> [Playlist] {
        load(from: playlistsFileURL) ?? []
    }

    func savePlaylists(_ playlists: [Playlist]) {
        save(playlists, to: playlistsFileURL)
    }

    // MARK: - Queue

    private var queueFileURL: URL { documentsDirectory.appendingPathComponent("queue.json") }

    func loadQueue() -> [PlaylistItem] {
        load(from: queueFileURL) ?? []
    }

    func saveQueue(_ items: [PlaylistItem]) {
        save(items, to: queueFileURL)
    }

    // MARK: - Video Playback Positions

    private var positionsFileURL: URL { documentsDirectory.appendingPathComponent("positions.json") }

    func loadPositions() -> [String: Double] {
        load(from: positionsFileURL) ?? [:]
    }

    func savePosition(_ position: Double, for videoURL: String) {
        var positions = loadPositions()
        positions[videoURL] = position
        save(positions, to: positionsFileURL)
    }

    func getPosition(for videoURL: String) -> Double? {
        loadPositions()[videoURL]
    }

    // MARK: - Recent Devices

    private var devicesFileURL: URL { documentsDirectory.appendingPathComponent("recent_devices.json") }

    func loadRecentDevices() -> [CastDevice] {
        load(from: devicesFileURL) ?? []
    }

    func saveRecentDevice(_ device: CastDevice) {
        var devices = loadRecentDevices()
        devices.removeAll { $0.host == device.host && $0.port == device.port }
        var updated = device
        updated.lastConnected = Date()
        devices.insert(updated, at: 0)
        if devices.count > 20 { devices = Array(devices.prefix(20)) }
        save(devices, to: devicesFileURL)
    }

    // MARK: - Per-site Ad Block Toggle

    private var siteAdBlockFileURL: URL { documentsDirectory.appendingPathComponent("site_adblock.json") }

    func loadSiteAdBlockSettings() -> [String: Bool] {
        load(from: siteAdBlockFileURL) ?? [:]
    }

    func setSiteAdBlock(domain: String, enabled: Bool) {
        var settings = loadSiteAdBlockSettings()
        settings[domain] = enabled
        save(settings, to: siteAdBlockFileURL)
    }

    func isAdBlockEnabled(for domain: String) -> Bool {
        let settings = loadSiteAdBlockSettings()
        return settings[domain] ?? AppSettings.shared.adBlockerEnabled
    }

    // MARK: - Generic Helpers

    private func load<T: Decodable>(from url: URL) -> T? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? decoder.decode(T.self, from: data)
    }

    private func save<T: Encodable>(_ object: T, to url: URL) {
        guard let data = try? encoder.encode(object) else { return }
        try? data.write(to: url, options: .atomic)
    }
}

File: WebCaster/src/services/storage/PersistenceController.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { HistoryEntry, Bookmark, PlaylistItem, Playlist, CastDevice, AppSettings, DEFAULT_SETTINGS } from '../../models/types';

const KEYS = {
  HISTORY: 'webcaster_history',
  BOOKMARKS: 'webcaster_bookmarks',
  QUEUE: 'webcaster_queue',
  PLAYLISTS: 'webcaster_playlists',
  SETTINGS: 'webcaster_settings',
  POSITIONS: 'webcaster_positions',
  RECENT_DEVICES: 'webcaster_recent_devices',
};

async function load<T>(key: string, fallback: T): Promise<T> {
  try {
    const raw = await AsyncStorage.getItem(key);
    return raw ? JSON.parse(raw) : fallback;
  } catch {
    return fallback;
  }
}

async function save(key: string, value: unknown): Promise<void> {
  try {
    await AsyncStorage.setItem(key, JSON.stringify(value));
  } catch (e) {
    console.warn('[Storage] Save failed:', key, e);
  }
}

export const Storage = {
  // History
  async loadHistory(): Promise<HistoryEntry[]> {
    return load(KEYS.HISTORY, []);
  },
  async saveHistory(entries: HistoryEntry[]): Promise<void> {
    await save(KEYS.HISTORY, entries);
  },
  async addHistoryEntry(entry: HistoryEntry): Promise<HistoryEntry[]> {
    const history = await this.loadHistory();
    const filtered = history.filter(h => h.url !== entry.url);
    const updated = [entry, ...filtered].slice(0, 500);
    await this.saveHistory(updated);
    return updated;
  },
  async clearHistory(): Promise<void> {
    await AsyncStorage.removeItem(KEYS.HISTORY);
  },

  // Bookmarks
  async loadBookmarks(): Promise<Bookmark[]> {
    return load(KEYS.BOOKMARKS, []);
  },
  async saveBookmarks(bookmarks: Bookmark[]): Promise<void> {
    await save(KEYS.BOOKMARKS, bookmarks);
  },
  async addBookmark(bookmark: Bookmark): Promise<Bookmark[]> {
    const bookmarks = await this.loadBookmarks();
    if (bookmarks.some(b => b.url === bookmark.url)) return bookmarks;
    const updated = [...bookmarks, bookmark];
    await this.saveBookmarks(updated);
    return updated;
  },
  async removeBookmark(id: string): Promise<Bookmark[]> {
    const bookmarks = await this.loadBookmarks();
    const updated = bookmarks.filter(b => b.id !== id);
    await this.saveBookmarks(updated);
    return updated;
  },

  // Queue
  async loadQueue(): Promise<PlaylistItem[]> {
    return load(KEYS.QUEUE, []);
  },
  async saveQueue(queue: PlaylistItem[]): Promise<void> {
    await save(KEYS.QUEUE, queue);
  },

  // Playlists
  async loadPlaylists(): Promise<Playlist[]> {
    return load(KEYS.PLAYLISTS, []);
  },
  async savePlaylists(playlists: Playlist[]): Promise<void> {
    await save(KEYS.PLAYLISTS, playlists);
  },

  // Playback positions
  async getPosition(videoURL: string): Promise<number | null> {
    const positions = await load<Record<string, number>>(KEYS.POSITIONS, {});
    return positions[videoURL] ?? null;
  },
  async savePosition(videoURL: string, position: number): Promise<void> {
    const positions = await load<Record<string, number>>(KEYS.POSITIONS, {});
    positions[videoURL] = position;
    await save(KEYS.POSITIONS, positions);
  },

  // Settings
  async loadSettings(): Promise<AppSettings> {
    return load(KEYS.SETTINGS, DEFAULT_SETTINGS);
  },
  async saveSettings(settings: AppSettings): Promise<void> {
    await save(KEYS.SETTINGS, settings);
  },

  // Recent devices
  async loadRecentDevices(): Promise<CastDevice[]> {
    return load(KEYS.RECENT_DEVICES, []);
  },
  async saveRecentDevice(device: CastDevice): Promise<void> {
    const devices = await this.loadRecentDevices();
    const filtered = devices.filter(d => !(d.host === device.host && d.port === device.port));
    const updated = [device, ...filtered].slice(0, 10);
    await save(KEYS.RECENT_DEVICES, updated);
  },

  // Clear all
  async clearAll(): Promise<void> {
    await AsyncStorage.multiRemove(Object.values(KEYS));
  },
};
